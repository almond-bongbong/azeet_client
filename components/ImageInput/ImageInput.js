import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import loadImage from 'blueimp-load-image';
import styled from 'styled-components';
import { hidden, blue } from 'style/mixin';

const ImageInputStyle = styled.div`
  & input {
    ${hidden}
  }
  
  & label {
    display: inline-block;
    padding: 10px 15px;
    border-radius: 2px;
    background-color: ${blue};
    color: #fff;
    cursor: pointer;
    transition: box-shadow .2s;
    
    &:hover {
      box-shadow: 1px 1px 3px rgba(0,0,0,0.2)
    }
    &:active {
      box-shadow: none;
    }
  }
`;

const imageLoadOptions = {
  maxWidth: 1200,
  orientation: true,
};

const ImageInput = ({ id, onChange }) => {
  const handleImage = useCallback((e) => {
    const { files } = e.target;
    const images = [];

    Array.from(files).forEach((file) => {
      loadImage(file, (img, meta) => {
        const { name, type, lastModified } = file;

        img.toBlob((blob) => {
          const convertedFile = new File([blob], name, { type, lastModified });

          images.push({ file: convertedFile, meta: meta.exif.getAll() });
          if (images.length === files.length && onChange instanceof Function) onChange(images);
        }, type);
      }, imageLoadOptions);
    });
  }, [onChange]);

  return (
    <ImageInputStyle>
      <label htmlFor={id}>
        <input
          id={id}
          type="file"
          onChange={handleImage}
          multiple
          accept="image/*"
          title="이미지 업로드"
        />
        <span>이미지 선택</span>
      </label>
    </ImageInputStyle>
  );
};

ImageInput.propTypes = {
  id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  onChange: PropTypes.func,
};

ImageInput.defaultProps = {
  id: 'image_input',
  onChange: null,
};

export default ImageInput;
