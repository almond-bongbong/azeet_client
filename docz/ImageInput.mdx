---
name: ImageInput
menu: Image
---

import { useState } from 'react';
import { Playground } from 'docz'
import ImageInput from 'components/ImageInput/ImageInput'

# ImageInput

## Basic usage

<Playground>
  {() => {
    const Test = () => {
      const [images, setImages] = useState([]);
      const handleImages = (images) => {
        // console.log(images.length, images);
        setImages(images);
      };

      return (
        <>
          <ImageInput onChange={handleImages} />
          <br /><br />
          {images.map(({ file, meta }) => (
            <div key={meta.DateTime}>
              <img src={URL.createObjectURL(file)} style={{ width: '100%' }} />
              <pre>{JSON.stringify(meta, null, 4)}</pre>
            </div>
          ))}
        </>
      );
    }

    return <Test />;
  }}
</Playground>
