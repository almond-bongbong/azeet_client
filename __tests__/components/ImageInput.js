import React from 'react';
import { render } from '@testing-library/react';
import ImageInput from 'components/ImageInput/ImageInput';

describe('Button', () => {
  it('matches snapshot', () => {
    const utils = render(<ImageInput />);
    expect(utils.container).toMatchSnapshot();
  });

  it('calls onChange File', async () => {
    const handleFile = jest.fn();
    const { getByTitle } = render(<ImageInput onChange={handleFile} />);
    getByTitle('이미지 업로드');
  });
});
