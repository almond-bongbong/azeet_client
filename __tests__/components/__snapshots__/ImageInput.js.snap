// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<div>
  <div
    class="ImageInput__ImageInputStyle-sc-69wo9x-0 gbPVZF"
  >
    <label
      for="image_input"
    >
      <input
        accept="image/*"
        id="image_input"
        multiple=""
        title="이미지 업로드"
        type="file"
      />
      <span>
        이미지 선택
      </span>
    </label>
  </div>
</div>
`;
